Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/ContentPlacementAnalyzer.cs

```diff
@@ -0,0 +1,532 @@
+                    // if we have less that the size, then take all
+                    if (extensionEntry.Value < weightedCount)
+                    {
+                        foreach (var a in m_artifactsPerExtension[extensionEntry.Key])
+                        {
+                            fileArtifacts[a] = m_fileContentMap[a];
+                            if (fileArtifacts.Count >= m_sampleSizeHardLimit)
+                            {
+                                return fileArtifacts;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // now, take a random set of files for each extension
+                        var randomSet = new HashSet<int>();
+                        while (randomSet.Count < weightedCount)
+                        {
+                            randomSet.Add(random.Next(extensionEntry.Value));
+                        }
+                        foreach (int position in randomSet)
+                        {
+                            var randomArtifact = m_artifactsPerExtension[extensionEntry.Key].ElementAt(position);
+                            fileArtifacts[randomArtifact] = m_fileContentMap[randomArtifact];
+                            if (fileArtifacts.Count >= m_sampleSizeHardLimit)
+                            {
+                                return fileArtifacts;
+                            }
+                        }
+                    }
+
+                }
+                return fileArtifacts;
+            }
+
+        }
+
+        public void WriteOutput(BXLArtifact artifact)
+        {
+            m_outputWriter.WriteStartObject();
+            // so, we are going to write a json here. The first line hast the main info
+            // (hash, build, queue, etc...). The following have only pip info
+            WriteJsonPropertyToStream("Hash", artifact.Hash);
+            WriteJsonPropertyToStream("BuildId", artifact.BuidId);
+            WriteJsonPropertyToStream("BuildQueue", artifact.BuildQueue);
+            WriteJsonPropertyToStream("BuildStartTimeTicks", artifact.BuildStartTimeTicks);
+            WriteJsonPropertyToStream("BuildDurationMs", artifact.BuildDurationMs);
+            WriteJsonPropertyToStream("ReportedFile", artifact.ReportedFile);
+            WriteJsonPropertyToStream("ReportedSize", artifact.ReportedSize);
+            WriteJsonPropertyToStream("TotalPips", artifact.TotalPips);
+            WriteJsonPropertyToStream("NumInputPips", artifact.NumInputPips);
+            WriteJsonPropertyToStream("NumOutputPips", artifact.NumOutputPips);
+            m_outputWriter.WritePropertyName("InputPips");
+            m_outputWriter.WriteStartArray();
+            for (int i = 0; i < artifact.NumInputPips; ++i)
+            {
+                m_outputWriter.WriteStartObject();
+                WriteJsonPropertyToStream("SemiStableHash", artifact.InputPips[i].SemiStableHash);

```

Comment:
>artifact.InputPips[i] [](start = 60, length = 21)

would you mind creating a variable for this? i.e.

`var inputPip = artifact.InputPips[i]`

and then use 
```
WriteJsonPropertyToStream("SemiStableHash", inputPip.SemiStableHash);
```

this should be a bit easier to read as well as faster. C# doesn't do inlining of this like C++ does since it doesn't fully check if the property and the indexer are side effect free so each call it will re-evaluate.

Reply:
note same for outputPip below.

---
In reply to: [314073692](https://github.com/microsoft/BuildXL/pull/734#discussion_r314073692) [](ancestors = 314073692)